eppo_core/src/obfuscation.rs

```diff
@@ -0,0 +1,69 @@
+use serde::{Deserialize, Serialize};
+use serde_with::serde_as;
+
+use crate::{ufc::ValueWire, Str};
+
+/// `md5::Digest` that implements `Serialize` and `Deserialize` (by converting to hex-encoded
+/// string).
+#[serde_as]
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Hash)]
+pub(crate) struct Md5HashedStr(#[serde_as(as = "serde_with::hex::Hex")] [u8; 16]);
+
+impl Md5HashedStr {
+    pub fn new(salt: &[u8], input: &[u8]) -> Md5HashedStr {
+        let mut ctx = md5::Context::new();
+        ctx.consume(salt);
+        ctx.consume(input);
+        ctx.compute().into()
+    }
+}
+
+impl From<md5::Digest> for Md5HashedStr {
+    fn from(value: md5::Digest) -> Self {
+        Md5HashedStr(value.0)
+    }
+}
+
+/// Same as [`Str`] but serializes as base64-encoded string.
+#[serde_as]
+#[derive(
+    Debug,
+    PartialEq,
+    Eq,
+    Hash,
+    PartialOrd,
+    Ord,
+    Clone,
+    Serialize,
+    Deserialize,
+    derive_more::From,
+    derive_more::Into,
+)]
+pub(crate) struct Base64Str(#[serde_as(as = "serde_with::base64::Base64")] pub(crate) Str);
+
+impl From<ValueWire> for Base64Str {
+    /// Convert value to base64. Booleans and numbers are converted to string first.
+    fn from(value: ValueWire) -> Base64Str {
+        let s = match value {
+            ValueWire::Boolean(b) => Str::from_static_str(if b { "true" } else { "false" }),
+            ValueWire::Number(n) => n.to_string().into(),
+            ValueWire::String(s) => s,
+        };
+        Base64Str(s)
+    }
+}
```

Comment:
just want to say that i am finding this rust pattern of 10 implementing desired traits and 2) pattern matching, very elegant. I was a little worried that rust would be like c++ and golang and not particularly "fun" to write but so far it's my top 3 favorite language.